### PR TITLE
Android UI updates

### LINKED ISSUE #1002: UI Improvements for Discovery
Epic: https://github.com/openmsupply/open-msupply/issues/996

UI improvements:

* If discovered site is not initialised we just show `Initialise`, this could be better phrased
* The back link (back to discovery) was just hacked in the UI, further thought 

## Further thoughts ##

Ideally, when Android app starts for the first time, we should be able to pick from Server/Client. If `server` is picked, server is discoverable and we connect straight to localhost, if `client` is picked, server is not discoverable and we go to discovery. Still should be able to navigate back to `discovery` from login/initialise screen if `client` was picked. And maybe in Admin windows can change from client to server ¯\_(ツ)_/¯

I think if we we have lots of Android client/server on the same network, it would be hard to find the correct one (don't even think the list in discovery is scrollable)

### PR DESCRIPTION
**Pull request by mark-prins**
Closes #1002 

Changes in this PR - I've scratched a few itches, so the PR has grown somewhat. 

- The initialise window has been replaced, so that a loading indicator can be shown. Particularly noticeable when you open the app, and have connected previously to a server. In the current app there's a bit of screen hopping, when the discovery page shows and then redirects to the app or login. Now a loader shows while you wait
- The term `Initialise` has been replaced with `[ Not configured ]`
- The UI for the list of servers has been tidied up
- There's now a refresh button for the server list ( which is actually pointless, but might make users feel better. The list is continuously updated, so the refresh only really clears the list 🤷 )
- If you cannot connect to a previously selected server, then an error is now shown
- The back link is changed to an 'edit' link
- Back link and server details moved to lower right, where the version info is shown
- There's the option to select server or client when you first open the app on android
- Android dnssd advertising only in server mode

Some examples

**Initial android screen**

Screenshot 2023-01-23 at 3 05 13 PM

**Server selection list**

Screenshot 2023-01-23 at 9 39 11 AM

**Server info and back to disco link**

Screenshot 2023-01-23 at 11 04 38 AM

**Client / Server option in admin**
image

**Loading indicator when re-opening the app**

https://user-images.githubusercontent.com/9192912/214154086-348d55a1-c913-40ff-aed0-8ba03c957acf.mp4

This is the workflow of the new page - it should show the loading indicator while the various decision paths are followed:

[](https://mermaid.live/edit#pako:eNptkc1qwzAQhF9l0cmB5AV8aEnjpkn6c0mhFHzZyBtbIEtmJScE2-9eWcqhlPpkduebHTGDkLYikYuasWvgsygNhG89yJ6ZjIc2rCdYrR5G02s9wlO2aax1FBfQYU2LeQvrOxilG60CO0Ix7NBBx3RRtnfgiC_E02_lMY5G2GZHj-yhUk7aMLnBiS1WEp1fQAK28c5LtrHGkPTgLWgrUaf7u-zN1sqkRAko4oVvciPssy9UHq6N0gSeb8rUMy-T1V2_j0aHIWUCvKDSeNL0GN8PyWmXtIfo_WFHeM3uekc6eCmbMuRwVb4BYrYMLTn3N9fMPv_PBp1Yipa4RVWFboaZK4VvqKVS5OG3ojP22peiNFOQYu_t8WakyD33tBR9V6GnQmFotRX5GbULU6qUt_ye-o61Tz_C8Kaw)

### CODE REVIEW

File: client/packages/common/src/hooks/useNativeClient/useNativeClient.ts [hunk 1]
```diff
@@ -97,26 +122,45 @@ export const useNativeClient = ({
...
-      setState(state => {
-        return {
-          ...state,
-          servers: uniqWith([...state.servers, ...servers], matchUniqueServer),
-          discoveryTimedOut: false,
-        };
-      });
-
-      clearTimeout(timeoutTimer);
+      if (servers.length === 0) return;
```
Comment (andreievg) [inline client/packages/common/src/hooks/useNativeClient/useNativeClient.ts:147]: Ohh yes, this must be why timeout wasn't working, thanks

File: client/packages/common/src/hooks/useNativeClient/useNativeClient.ts [hunk 2]
```diff
@@ -126,11 +170,15 @@ export const useNativeClient = ({
     if (!nativeAPI) return;
     if (!autoconnect) return;
     if (!previousServer) return;
+    if (mode !== NativeMode.Client) return;
```
Comment (andreievg) [inline client/packages/common/src/hooks/useNativeClient/useNativeClient.ts:173]: does this mean if `Server` mode was picked, it wouldn't also work as client ?

Comment (mark-prins) [reply client/packages/common/src/hooks/useNativeClient/useNativeClient.ts:173]: I was catering for a pretty specific scenario, which, yeah, affects the autoconnect. 
Will just set the previousServer when selecting server

**Review by andreievg — CHANGES_REQUESTED** [review]
Overall this is great! thank you, UI looks awesome and hook changes seem solid and quite readable. 

Had some issues, as per suggestions.  And had a problem where I just couldn't discover local server on emulator (it just suddenly stopped working), had to restart emulator, and seems to work fine after that

**Comment by mark-prins** [comment]
Changes made - still issues with discovery that I'm looking into.

- there's some reliability and repeatability problems
- the `isLocal` prop isn't correct
- the message 'cannot find local server' shows, just after showing a local server
 
as an example, on the tablet with IP `192.168.178.91` these are some of the results:

Screenshot 2023-01-31 at 9 59 57 AM

Screenshot 2023-01-31 at 9 50 11 AM

Screenshot 2023-01-31 at 10 01 28 AM

File: client/packages/android/app/src/main/java/org/openmsupply/client/NativeApi.java [hunk 3]
```diff
@@ -60,7 +62,11 @@ public void goBackToDiscovery(PluginCall call) {
     }
 
     // Advertise local remote server on network
-    void advertiseService() {
+    @PluginMethod()
+    public void advertiseService() {
```
[the remark this answers is not in the document]
Comment (andreievg) [reply client/packages/android/app/src/main/java/org/openmsupply/client/NativeApi.java:66]: I think this still needs to be changed

**Comment by mark-prins** [comment]
Updated now - have fixed some issues with the interval & timers. all working ok afaik

**Comment by andreievg** [comment]
@mark-prins it looks like `resolveService` really needs to be called in synchronous way, otherwise hardware_id (and maybe other information) is being overwritten. Although it's probably not 100 thread safe, this seems to work:  https://github.com/openmsupply/open-msupply/compare/Server-discovery-resolution

How I tried to replicate the issue:

On develop branch, had two tablets and local server running. Added log to `onServiceFound` and to `onServiceResolved`. On restart most time hardware_id was mixed up, tried something with the boolean to just run resolveService when one is not running, this worked but was missing some services that were `onServiceFound`, so added a queue and this seems to work and quite fast.

**Comment by mark-prins** [comment]
added the latest suggestion - working nicely and much faster!

File: client/packages/android/app/src/main/java/org/openmsupply/client/NativeApi.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mark-prins) [reply client/packages/android/app/src/main/java/org/openmsupply/client/NativeApi.java]: 🤔  It's showing correctly for me on line `75`

**Review by andreievg — APPROVED** [review]
Thank you @mark-prins!

**Pull request merged**